### PR TITLE
feat(p2d): move several settings into YAML

### LINKED ISSUE #3: P2D: Move several settings into YAML
ref: https://github.com/DOMjudge/domjudge/pull/1101

Move

- Problem Name
- Memory Limit
- Output Limit (do we have to determine it?)
- Source (from the polygon problem url?)

into the yaml.

The parser of ini is dangerous.

### PR DESCRIPTION
**Pull request by cubercsl**
fix #3

### CODE REVIEW

File: p2d/bin/p2d [hunk 1]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+export PYTHONPATH="$(dirname "$(dirname "$(readlink -f "$0")")"):$PYTHONPATH"
+export CONFIG_PATH="$(dirname "$(dirname "$(readlink -f "$0")")")"
+export TESTLIB_PATH=$CONFIG_PATH
+export LOG_PATH='/var/log/p2d'
```
Comment (cubercsl) [inline p2d/bin/p2d]: 用户没有写入权限，应该找个 userspace 的路径。

File: p2d/p2d.py [hunk 2]
```diff
@@ -40,29 +39,34 @@ def parse_problem():
...
+            f'color = {color}'
         ]
         [*map(logger.info, ini_content)]
         with open(ini_file, 'w', encoding='utf-8') as f:
             f.writelines(map(lambda s: s + '\n', ini_content))
         logger.info(END_OF_SUBPROCESS)
 
-    def add_output_validator():
+    def write_yaml(name, memlimit):
+        logger.info('Add \'priblem.yaml\':')
```
Comment (yang-er) [inline p2d/p2d.py]: What is `priblem.yaml`

Comment (cubercsl) [reply p2d/p2d.py]: cao

File: p2d/bin/p2d [hunk 3]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+export PYTHONPATH="$(dirname "$(dirname "$(readlink -f "$0")")"):$PYTHONPATH"
+export CONFIG_PATH="$(dirname "$(dirname "$(readlink -f "$0")")")"
+export TESTLIB_PATH=$CONFIG_PATH
+export LOG_PATH='~/.p2d'
+exec python -m p2d "$@"
```
Comment (yang-er) [inline p2d/bin/p2d]: Seems like `python3` should be used. My Ubuntu 20.04 WSL doesn't contains `python`, and my Ubuntu 18.04 VM instances defaults to py2.

File: p2d/bin/p2d [hunk 4]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+export PYTHONPATH="$(dirname "$(dirname "$(readlink -f "$0")")"):$PYTHONPATH"
+export CONFIG_PATH="$(dirname "$(dirname "$(readlink -f "$0")")")"
+export TESTLIB_PATH=$CONFIG_PATH
+export LOG_PATH='~/.p2d'
```
Comment (yang-er) [inline p2d/bin/p2d]: `~/.p2d` creates an directory called `~`......

I think `$HOME` is better?

Comment (yang-er) [reply p2d/bin/p2d]:
```
export LOG_PATH="$HOME/.p2d"
```

File: p2d/p2d.py [hunk 5]
```diff
@@ -29,40 +28,45 @@ def main(args):
...
 
     def parse_problem():
         logger.info('Parse \'problem.xml\':')
-        xml_file = '{}/problem.xml'.format(package_dir)
+        xml_file = f'{package_dir}/problem.xml'
         root = xml.etree.ElementTree.parse(xml_file)
         name = root.find('names').find('name').attrib['value']
-        timelimit = float(math.ceil(float(root.find('judging').find('testset').find('time-limit').text) / 1000.0))
+        timelimit = int(root.find('judging').find('testset').find('time-limit').text) / 1000.0
+        memlimit = int(root.find('judging').find('testset').find('memory-limit').text) // 1024
```
Comment (yang-er) [inline p2d/p2d.py]: According to 

https://github.com/DOMjudge/domjudge/blob/acf58a7e3fe0c6c47c860490b280a42ba7ab76de/webapp/src/Service/ImportProblemService.php#L378

It should be megabytes.

This should be `// 1024 // 1024` or `// 1048576`

**Comment by yang-er** [comment]
image

Should we put `testlib.h` into the archive?

File: p2d/bin/p2d (hunk 3 above)
Comment (yang-er) [reply p2d/bin/p2d]: @xzm2000 Do you have any suggestions? Something like `/usr/bin/env python3`?

File: p2d/p2d.py [hunk 6]
```diff
@@ -114,55 +119,59 @@ def get_checker_md5(checker):
             if interactor is not None:
                 logger.info('Use custom interactor.')
                 with open(yaml_file, 'w', encoding='utf-8') as f:
+                    f.write(yaml_content)
                     f.write('validation: custom interactive\n')
                 ensure_dir(interactor_dir)
-                copyfile(testlib, '{}/testlib.h'.format(interactor_dir))
-                interactor_file = '{}/{}'.format(package_dir, interactor.find('source').attrib['path'])
-                copyfile(interactor_file, '{}/interactor.cpp'.format(interactor_dir))
+                copyfile(testlib, f'{interactor_dir}/testlib.h', follow_symlinks=False)
```
Comment (yang-er) [inline p2d/p2d.py]: image

When you are not in the p2d directory, `../testlib.h` may not exist. So copyfile is failed and the warning is ignored. I think the symbolic link is not acceptable.

**Review by yang-er — COMMENTED** [review]
> image
> 
> Should we put `testlib.h` into the archive?

This is the reason missing.

File: p2d/p2d.py [hunk 7]
```diff
@@ -114,55 +119,59 @@ def get_checker_md5(checker):
...
             elif checker is not None:
                 logger.info('Use custom checker.')
                 with open(yaml_file, 'w', encoding='utf-8') as f:
+                    f.write(yaml_content)
                     f.write('validation: custom\n')
                 ensure_dir(checker_dir)
-                copyfile(testlib, '{}/testlib.h'.format(checker_dir))
-                checker_file = '{}/{}'.format(package_dir, checker.find('source').attrib['path'])
-                copyfile(checker_file, '{}/checker.cpp'.format(checker_dir))
+                copyfile(testlib, f'{checker_dir}/testlib.h', follow_symlinks=False)
```
Comment (yang-er) [inline p2d/p2d.py]: What about marking `TESTLIB_PATH` as `"$CONFIG_PATH"/../` or `$(dirname "$CONFIG_PATH")`?

**Review by yang-er — APPROVED** [review]

File: p2d/p2d.py (hunk 7 above)
Comment (cubercsl) [reply p2d/p2d.py]: have changed in python now~

**Review by yang-er — APPROVED** [review]

**Pull request merged**